src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,401 @@
+            MirRelationExpr::Threshold { input } => {
+                // Threshold requires all columns, as collapsing any distinct values
+                // has the potential to change how it thresholds counts. This could
+                // be improved with reasoning about distinctness or non-negativity.
+                let arity = input.arity();
+                self.action(input, (0..arity).collect(), gets);
+                (0..arity).collect()
+            }
+            MirRelationExpr::ArrangeBy { input, keys } => {
+                // If a key set contains keys that are not demanded upstream,
+                // delete the key set.
+                // If `keys` becomes empty, delete the arrangeby, and keep
+                // pushing the projection. Otherwise, do not push the project past
+                // the ArrangeBy
+                for i in (0..keys.len()).rev() {
+                    if keys[i]
+                        .iter()
+                        .any(|k| k.support().iter().any(|c| !desired_projection.contains(c)))
+                    {
+                        keys.remove(i);
+                    }
+                }
+                if keys.is_empty() {
+                    self.action(input, desired_projection.clone(), gets);
+                    *relation = input.take_dangerous();
+                    desired_projection.clone()
+                } else {
+                    let arity = input.arity();
+                    self.action(input, (0..arity).collect(), gets);
+                    (0..arity).collect()
+                }
+            }
+            MirRelationExpr::DeclareKeys { input, keys } => {
+                // TODO[btv] - If and when we add a "debug mode" that asserts whether this is truly a key,
+                // we will probably need to add the key to the set of demanded
+                // columns.
+
+                // Current behavior is that if a key is not contained with the
+                // desired_projection, then it is not relevant to the query plan
+                // and can be removed.
+                for i in (0..keys.len()).rev() {
+                    if keys[i].iter().any(|k| !desired_projection.contains(k)) {
+                        keys.remove(i);
+                    }
+                }
+                self.action(input, desired_projection.clone(), gets);
+                if keys.is_empty() {
+                    *relation = input.take_dangerous();
+                }
+                desired_projection.clone()
+            }
+        };
+        let add_project = desired_projection != actual_projection;
+        if add_project {
+            reverse_permute_columns(desired_projection.iter_mut(), actual_projection.iter());
+            *relation = relation.take_dangerous().project(desired_projection);
+        }
+    }
+}
+
+fn get_sorted_vec(set: HashSet<usize>) -> Vec<usize> {
+    let mut result = set.into_iter().collect::<Vec<usize>>();
+    result.sort();
+    result
+}
```

Comment:
If this happens a lot, maybe we should just use a btree? I have a minor bias towards them to ensure determinism of enumeration order.